src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -0,0 +1,251 @@
+
+            this.CheckSignature(header, prevUtxo);
+
+            this.CheckStakeKernelHash((PosRuleContext)context, prevUtxo, header, chainedHeader);
+
+            this.CheckCoinstakeMerkleProof(header);
+
+            this.CheckHeaderSignatureWithCoinstakeKernel(header, prevUtxo);
+        }
+
+        /// <summary>
+        /// Checks the coinstake to make sure it is not null.
+        /// </summary>
+        /// <param name="header">The proven blockheader.</param>
+        private void CheckCoinstakeIsNotNull(ProvenBlockHeader header)
+        {
+            if (header.Coinstake != null)
+                return;
+
+            this.Logger.LogTrace("(-)[COINSTAKE_IS_NULL]");
+            ConsensusErrors.EmptyCoinstake.Throw();
+        }
+
+        /// <summary>
+        /// Fetches and validates coins from coins view.
+        /// </summary>
+        /// <param name="header">The header.</param>
+        private FetchCoinsResponse GetAndValidateCoins(ProvenBlockHeader header)
+        {
+            // First try finding the previous transaction in database.
+            TxIn txIn = header.Coinstake.Inputs[0];
+            FetchCoinsResponse coins = this.PosParent.UtxoSet.FetchCoinsAsync(new[] { txIn.PrevOut.Hash }).GetAwaiter().GetResult();
+            if ((coins == null) || (coins.UnspentOutputs.Length != 1))
+                ConsensusErrors.ReadTxPrevFailed.Throw();
+
+            return coins;
+        }
+
+        /// <summary>
+        /// Gets and validates unspent outputs based of coins fetched from coin view.
+        /// </summary>
+        /// <param name="coins">The coins.</param>
+        private UnspentOutputs GetAndValidatePreviousUtxo(FetchCoinsResponse coins)
+        {
+            UnspentOutputs prevUtxo = coins.UnspentOutputs[0];
+            if (prevUtxo == null)
+            {
+                this.Logger.LogTrace("(-)[PREV_UTXO_IS_NULL]");
+                ConsensusErrors.ReadTxPrevFailed.Throw();
+            }
+
+            return prevUtxo;
+        }
+
+        /// <summary>
+        /// Checks if coinstake transaction is valid.
+        /// </summary>
+        /// <param name="header">The proven block header.</param>
+        private void CheckIfCoinstakeIsTrue(ProvenBlockHeader header)
+        {
+            if (header.Coinstake.IsCoinStake)
+                return;
+
+            this.Logger.LogTrace("(-)[PROVEN_HEADER_INVALID_MERKLE_PROOF_SIZE]");
+            ConsensusErrors.NonCoinstake.Throw();
+        }
+
+        /// <summary>
+        /// Checks whether header time is equal to the timestamp of the coinstake tx and if coinstake tx
+        /// timestamp is divisible by 16 (using timestamp mask).
+        /// </summary>
+        /// <param name="header">The proven block header.</param>
+        private void CheckHeaderAndCoinstakeTimes(ProvenBlockHeader header)
+        {
+            uint coinstakeTime = header.Coinstake.Time;
+            uint headerTime = header.Time; 
```

Comment:
trailing whitespaces.
Install Trailing Whitespace Visualizer to your VS, it removes trailing whitespaces on file save.